### ISSUE TITLE
Implement true parallelism in RuboCop

### ISSUE THREAD

**Issue by tdeo** [rubocop/rubocop#15075]
### Before submitting

- [x] I have searched the existing issues and this hasn't been requested yet.

### Is your feature request related to a problem?

When running RuboCop in parallel mode (which is the default, faster mode), it's frustrating to be in the dark while the actual work is happening during [warm_cache](https://github.com/rubocop/rubocop/blob/1fc18f7eff49084b2f700cd7ab33b4a2546292d2/lib/rubocop/runner.rb#L76) where no progress is being reported, and only get some output during the few seconds that `inspect_files` subsequently takes using a valid cache.

### Describe the solution you'd like

I'd like to get some progress output while the `warm_cache` step of the run is happening, since this is the one taking the longest, by far.

### Describe alternatives you've considered

I see 2 options:

1. The most straightforward to implement: set a first progress reporter during `warm_cache`, which is quite easy to implement using `parallel` gem's [progress API](https://github.com/grosser/parallel#progress--eta) - I haven't tested it the whole way but it looks like those changes would provide this functionality:

<details>
<summary>`warm_cache` with progress reporting</summary>

```ruby
    def warm_cache(target_files)
      formatter_set.started(target_files)
      saved_options = @options.dup
      if target_files.size <= 1
        puts 'Skipping parallel inspection: only a single file needs inspection' if @options[:debug]
        return
      end
      puts 'Running parallel inspection' if @options[:debug]
      %i[autocorrect safe_autocorrect].each { |opt| @options[opt] = false }
      Parallel.each(
        target_files,
        start: ->(file, _index) { file_started(file) },
        finish: ->(file, _index, _result) { file_finished(file, []) },
      ) do |target_file|
        file_offenses(target_file)
        nil
      end
    ensure
      @options = saved_options
      formatter_set.finished(target_files)
      formatter_set.close_output_files
    end
```
</details>

2. Make RuboCop parallel run truly parallel, without combining 2 sequential runs to first warm the cache in parallel, and then perform an actual run sequentially. This gets a bit trickier to implement but I've got a decent start with the following approach:

<details>
<summary>`RuboCop::Runner#inspect_files(files)` implementation with true parallelism</summary>

```ruby
    def inspect_files(files)
      formatter_set.started(files)

      file_iterator(files) do |file|
        offenses = process_file(file)
        if offenses.any? { |o| considered_failure?(o) && offense_displayed?(o) }
          raise Parallel::Break if @options[:fail_fast]

          next [offenses, false]
        end
        [offenses, true]
      end
    ensure
      formatter_set.finished(@inspected_files.freeze)
      formatter_set.close_output_files
    end

    def process_file(file)
      file_offenses(file)
    rescue InfiniteCorrectionLoop => e
      raise e if @options[:raise_cop_error]

      errors << e
      warn Rainbow(e.message).red
      e.offenses.compact.sort.freeze
    end

    def file_iterator(files, &block)
      all_passed = true
      on_start = ->(file, _index) { file_started(file) }
      on_finish = lambda do |file, _index, (offenses, passed)|
        file_finished(file, offenses)
        all_passed &&= passed
      end
      if @options[:parallel]
        Parallel.each(files, start: on_start, finish: on_finish, &block)
      else
        files.each do |file|
          file_started(file)
          offenses, passed = yield file
          file_finished(file, offenses)
          all_passed &&= passed
        rescue Parallel::Break
          break
        end
      end
      all_passed
    end
```
</details>

[This changeset](https://github.com/rubocop/rubocop/compare/master...tdeo:rubocop:true_runner_parallelism) right now has 7 failures left over RuboCop's test suite (and a couple internal_investigation issues with Complexity) which are about missing warnings and order of individual file output. 

Before spending further time into it, I was hoping to get opinions about the better direction and see if I should pursue option 2 or settle for option 1 (or just don't bother about this change)

**Comment by tdeo** [comment]
Hello @bbatsov @koic - would anyone have an opinion on the direction to pursue?
